### ISSUE TITLE
Docker plugin unable to pull image despite registry login details provided

### ISSUE THREAD

**Issue by pteichner** [jenkinsci/docker-plugin#528]
If you get some troubles with docker-plugin, please report
 - [ ] docker-plugin: 0.18
 - [ ] jenkins version you use: 2.73.2
 - [ ] docker engine version you use: 1.12.6
 - [ ] stack trace / logs / any technical details that could help diagnose this issue

```
Oct 17, 2017 9:52:29 AM com.nirima.jenkins.plugins.docker.DockerCloud$1 call
SEVERE: Error in provisioning; template='DockerTemplate{configVersion=2, labelString='rhel5-docker-python', launcher=com.nirima.jenkins.plugins.docker.launcher.DockerComputerSSHLauncher@58f93a21, remoteFsMapping='/var/lib/jenkins', remoteFs='/home/jenkins', instanceCap=4, mode=NORMAL, retentionStrategy=com.nirima.jenkins.plugins.docker.strategy.DockerOnceRetentionStrategy@3df2e3e6, numExecutors=1, dockerTemplateBase=DockerTemplateBase{image=testdocker.uex-repo-test.hq.k.grp/cd/nagra/centos-5-python-jenkinsagent:latest}, removeVolumes=true, pullStrategy=PULL_LATEST, nodeProperties=[]}' for cloud='build-docker-01'
com.github.dockerjava.api.exception.InternalServerErrorException: {"message":"Get https://testdocker.uex-repo-test.hq.k.grp/v2/cd/nagra/centos-5-python-jenkinsagent/manifests/latest: unknown: Authentication is required"}

        at com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:109)
        at com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:33)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:134)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:748)
```
![jenkins](https://user-images.githubusercontent.com/28305157/31658912-566eef16-b32a-11e7-810e-a3308769a911.png)

**Comment by ndeloof** [comment]
What's the nature of your registry server ?

**Comment by pteichner** [comment]
It's an Artifactory server - but I was under the impression they all use the same API defined by Docker itself. It does definitely work from docker cli.

**Comment by ndeloof** [comment]
I assume so as well, but wanted to know so I can try to reproduce :P

**Comment by ndeloof** [comment]
The CLI does lot's more but just pass credentials, it also manage token TTL and refresh, as well as credentials storage. Docker plugin uses a Java API client and as such need to directly give the adequate authentication data to the API.

**Comment by ndeloof** [comment]
Should be fixed by 577fa4838a6df1d4d310075b64a3755414c6017e
Please reopen if this issue persists in 1.0.1

**Comment by hypery2k** [comment]
I'm hitting the same error with 1.1.2:
```
com.github.dockerjava.api.exception.InternalServerErrorException: {"message":"Get https://docker-hub.holisticon.de:4443/v2/holi_ci/jenkins_node_jdk8/manifests/latest: no basic auth credentials"}

        at com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:109)
        at com.github.dockerjava.netty.handler.HttpResponseHandler.channelRead0(HttpResponseHandler.java:33)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:997)
        at io.netty.channel.epoll.EpollDomainSocketChannel$EpollDomainUnsafe.epollInReady(EpollDomainSocketChannel.java:138)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:401)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:306)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:748)

Feb 07, 2018 11:09:22 AM com.nirima.jenkins.plugins.docker.DockerCloud$1 run
SEVERE: Error in provisioning; template='DockerTemplate{configVersion=2, labelString='java jdk8', connector=io.jenkins.docker.connector.DockerComputerAttachConnector@3f7816a4, remoteFs='/tmp', instanceCap=2147483647, mode=NORMAL, retentionStrategy=com.nirima.jenkins.plugins.docker.strategy.DockerOnceRetentionStrategy@dcfc903, dockerTemplateBase=DockerTemplateBase{image='docker-hub.holisticon.de:4443/holi_ci/jenkins_node_jdk8', pullCredentialsId='', registry=DockerRegistryEndpoint[null;credentialsId=null], dockerCommand='', hostname='', dnsHosts=[], network='', volumes=[], volumesFrom2=[], environment=[], bindPorts='', bindAllPorts=false, memoryLimit=null, memorySwap=null, cpuShares=null, privileged=false, tty=false, macAddress='null', extraHosts=[]}, removeVolumes=false, pullStrategy=PULL_ALWAYS, nodeProperties=[]}' for cloud='docker'
com.github.dockerjava.api.exception.InternalServerErrorException: {"message":"Get https://docker-hub.holisticon.de:4443/v2/holi_ci/jenkins_node_jdk8/manifests/latest: no basic auth credentials"}

```

Whats strange is that for the template the registry is empty:
```
{  
   configVersion=2,
   labelString='java jdk8',
   connector=io.jenkins.docker.connector.DockerComputerAttachConnector@3f7816a4,
   remoteFs='/tmp',
   instanceCap=2147483647,
   mode=NORMAL,
   retentionStrategy=com.nirima.jenkins.plugins.docker.strategy.DockerOnceRetentionStrategy@dcfc903,
   dockerTemplateBase=DockerTemplateBase   {  
      image='docker-hub.holisticon.de:4443      /holi_ci/jenkins_node_jdk8',
      pullCredentialsId='',
      registry=DockerRegistryEndpoint      [  
         null         ;credentialsId=null
      ],
      dockerCommand='',
      hostname='',
      dnsHosts=      [  

      ],
      network='',
      volumes=      [  

      ],
      volumesFrom2=      [  

      ],
      environment=      [  

      ],
      bindPorts='',
      bindAllPorts=false,
      memoryLimit=null,
      memorySwap=null,
      cpuShares=null,
      privileged=false,
      tty=false,
      macAddress='null',
      extraHosts=      [  

      ]
   },
   removeVolumes=false,
   pullStrategy=PULL_ALWAYS,
   nodeProperties=   [  

   ]
}
```
